File: Mars-Images/MarsPhoto.m
//
//  MarsPhoto.m
//  Mars-Images
//
//  Created by Mark Powell on 12/22/13.
//  Copyright (c) 2013 Powellware. All rights reserved.
//

#import "MarsPhoto.h"
#import "CameraModel.h"
#import "Math.h"
#import "ImageUtility.h"
#import "MarsImageNotebook.h"
#import "SDWebImageManager.h"
#import "SDWebImageDecoder.h"

@interface MarsPhoto () {
    BOOL _anaglyphLoadingInProgress;
    id <SDWebImageOperation> _leftImageOperation;
    id <SDWebImageOperation> _rightImageOperation;
}
@end

@implementation MarsPhoto

double d1[3], d2[3];

- (id) initWithResource: (EDAMResource*) resource
                   note: (EDAMNote*) note
                    url: (NSURL*) url {
    self = [super initWithURL:url];
    _note = note;
    _resource = resource;
    _model_json = nil;
    self.caption = [[MarsImageNotebook instance].mission captionText:note];
    return self;
}

- (id) initAnaglyph: (NSArray*) leftAndRight
               note: (EDAMNote*) note {
    self = [super init];
    _note = note;
    _leftAndRight = leftAndRight;
    self.caption = [[MarsImageNotebook instance].mission captionText:note];
    return self;
}

- (BOOL) isGrayscale {
    if ([_note.title rangeOfString:@"Mastcam"].location != NSNotFound ||
        [_note.title rangeOfString:@"MAHLI"].location != NSNotFound ||
        [_note.title rangeOfString:@"MARDI"].location != NSNotFound ||
        [_note.title rangeOfString:@"Color"].location != NSNotFound)
        return NO;

    return _leftAndRight == nil;
}

- (BOOL) includedInMosaic {
    return ([_note.title rangeOfString:@"Navcam"].location != NSNotFound
//           || [_note.title rangeOfString:@"Mastcam"].location != NSNotFound
//           || [_note.title rangeOfString:@"Pancam"].location != NSNotFound
            );
    //TODO include these when texture memory can be managed effectively
}

- (NSURL*) url: (EDAMResource*) resource {
    NSString* resGUID = resource.guid;
    NSString* address = [NSString stringWithFormat:@"%@res/%@", Evernote.instance.user.webApiUrlPrefix, resGUID];
    return [NSURL URLWithString:address];
}

- (void)performLoadUnderlyingImageAndNotify {
    // Load async from web (using SDWebImage)
    if (_leftAndRight) {
        SDWebImageManager *manager = [SDWebImageManager sharedManager];
        @try {
            EDAMResource* leftResource = [_leftAndRight objectAtIndex:0];
            NSURL* leftUrl = [self url:leftResource];
            _leftImageOperation = [manager downloadImageWithURL:leftUrl
                                                      options:0
                                                 progress:^(NSInteger receivedSize, NSInteger expectedSize) {
                                                     if (expectedSize > 0) {
                                                         float progress = receivedSize / (float)expectedSize;
                                                         NSDictionary* dict = [NSDictionary dictionaryWithObjectsAndKeys:
                                                                               [NSNumber numberWithFloat:progress], @"progress",
                                                                               self, @"photo", nil];
                                                         [[NSNotificationCenter defaultCenter] postNotificationName:MWPHOTO_PROGRESS_NOTIFICATION object:dict];
                                                     }
                                                 }
                                                completed:^(UIImage *image, NSError *error, SDImageCacheType cacheType, BOOL finished, NSURL* url) {
                                                    if (error) {
                                                        NSLog(@"SDWebImage failed to download image: %@", error);
                                                    }
                                                    _leftImageOperation = nil;
                                                    _leftImage = image;
                                                    if (_rightImage) {
                                                        self.underlyingImage = [ImageUtility anaglyphImages:_leftImage right:_rightImage];
                                                        [self decompressImageAndFinishLoading];
                                                    }
                                                }];
        } @catch (NSException *e) {
            NSLog(@"Photo from web: %@", e);
            _leftImageOperation = nil;
            [self decompressImageAndFinishLoading];
        }

        @try {
            EDAMResource* rightResource = [_leftAndRight objectAtIndex:1];
            NSURL* rightUrl = [self url:rightResource];
            _rightImageOperation = [manager downloadImageWithURL:rightUrl
                                                   options:0
                                                  progress:^(NSInteger receivedSize, NSInteger expectedSize) {
                                                      if (expectedSize > 0) {
                                                          float progress = receivedSize / (float)expectedSize;
                                                          NSDictionary* dict = [NSDictionary dictionaryWithObjectsAndKeys:
                                                                                [NSNumber numberWithFloat:progress], @"progress",
                                                                                self, @"photo", nil];
                                                          [[NSNotificationCenter defaultCenter] postNotificationName:MWPHOTO_PROGRESS_NOTIFICATION object:dict];
                                                      }
                                                  }
                                                 completed:^(UIImage *image, NSError *error, SDImageCacheType cacheType, BOOL finished, NSURL* url) {
                                                     if (error) {
                                                         NSLog(@"SDWebImage failed to download image: %@", error);
                                                     }
                                                     _rightImageOperation = nil;
                                                     _rightImage = image;
                                                     if (_leftImage) {
                                                         self.underlyingImage = [ImageUtility anaglyphImages:_leftImage right:_rightImage];
                                                         [self decompressImageAndFinishLoading];
                                                     }
                                                 }];
        } @catch (NSException *e) {
            NSLog(@"Photo from web: %@", e);
            _rightImageOperation = nil;
            [self decompressImageAndFinishLoading];
        }
    }
    else {
        [super performLoadUnderlyingImageAndNotify];
    }
}

- (void)decompressImageAndFinishLoading {
    NSAssert([[NSThread currentThread] isMainThread], @"This method must be called on the main thread.");
    if (self.underlyingImage) {
        // Decode image async to avoid lagging when UIKit lazy loads
        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
            self.underlyingImage = [UIImage decodedImageWithImage:self.underlyingImage];
            dispatch_async(dispatch_get_main_queue(), ^{
                // Finish on main thread
                [self imageLoadingComplete];
            });
        });
    } else {
        // Failed
        [self imageLoadingComplete];
    }
}

- (void)imageLoadingComplete {
    NSAssert([[NSThread currentThread] isMainThread], @"This method must be called on the main thread.");
    // Complete so notify
    _anaglyphLoadingInProgress = NO;
    // Notify on next run loop
    [self performSelector:@selector(postCompleteNotification) withObject:nil afterDelay:0];
}

- (NSArray*) modelJson {
    if (!_model_json) {
        NSString* cmod_string = self.resource.attributes.cameraModel;
        if (!cmod_string || cmod_string.length == 0)
            return nil;
        NSData* json = [cmod_string dataUsingEncoding:NSUTF8StringEncoding];
        NSError* error;
        _model_json = [NSJSONSerialization JSONObjectWithData:json options:nil error:&error];
    }
    return _model_json;
}

- (double) angularDistance: (MarsPhoto*) otherImage {
    NSArray* v1 = [CameraModel pointingVector:_model_json];
    NSArray* v2 = [CameraModel pointingVector:[otherImage modelJson]];
    if (v1.count==0 || v2.count==0)
        return 0;
    
    d1[0] = [(NSNumber*)[v1 objectAtIndex:0] doubleValue];
    d1[1] = [(NSNumber*)[v1 objectAtIndex:1] doubleValue];
    d1[2] = [(NSNumber*)[v1 objectAtIndex:2] doubleValue];
    d2[0] = [(NSNumber*)[v2 objectAtIndex:0] doubleValue];
    d2[1] = [(NSNumber*)[v2 objectAtIndex:1] doubleValue];
    d2[2] = [(NSNumber*)[v2 objectAtIndex:2] doubleValue];
    
    double dot = [Math dot:d1 b:d2];
    return acos(dot);
}

@end

File: MarsImagesIOS/MarsPhoto.swift
//
//  MarsPhoto.swift
//  MarsImagesIOS
//
//  Created by Mark Powell on 7/28/17.
//  Copyright © 2017 Mark Powell. All rights reserved.
//

import Foundation
import POWImageGallery
import SDWebImage
import SwiftyJSON

class MarsPhoto: ImageCreator {
    
    var imageset:Imageset
    var indexInImageset:Int
    var sourceUrl:String
    var modelJson:JSON?
    var caption:String
    
    var leftImage:UIImage?
    var rightImage:UIImage?
    var leftAndRight:(String,String)?
    var isIncludedInMosaic:Bool {
        return imageset.title.range(of:"Navcam") != nil
            || imageset.title.range(of:"Mastcam Left") != nil
            || imageset.title.range(of:"Pancam") != nil
    }
    
    required public init (url:URL, imageset: Imageset, indexInImageset: Int, sourceUrl:String, modelJsonString:String?) {
        self.imageset = imageset
        self.indexInImageset = indexInImageset
        self.sourceUrl = sourceUrl
        if let model = modelJsonString {
            modelJson = JSON(parseJSON:model)
        }
        self.caption = Mission.currentMission().caption(self.imageset.title)
        super.init(url:url, delegate:nil)
    }
    
    init (_ imageset: Imageset, leftAndRight:(String,String)) {
        self.imageset = imageset
        self.leftAndRight = leftAndRight
        self.sourceUrl = ""
        self.indexInImageset = 0
        self.caption = Mission.currentMission().caption(self.imageset.title)
        super.init(url:URL(string:leftAndRight.0)!, delegate:nil)
    }
    
    override open func requestImage() {
        guard (leftAndRight != nil) else {
            super.requestImage()
            return
        }

        if loadInProgress {
            return
        }
        loadInProgress = true

        let leftUrl = URL(string:leftAndRight!.0)
        SDWebImageManager.shared().loadImage(with: leftUrl, options: .refreshCached,
                                             progress:  { (receivedSize, expectedSize, targetUrl) -> Void in
                                                self.delegate?.progress?(receivedSize:receivedSize, expectedSize:expectedSize)
        },
                                             completed: { (image, data, error, cacheType, finished, imageURL) -> Void in
                                                if let image = image {
                                                    self.leftImage = image
                                                    self.processImage()
                                                } else {
                                                    self.delegate?.failure()
                                                }
        })
        let rightUrl = URL(string:leftAndRight!.1)
        SDWebImageManager.shared().loadImage(with: rightUrl, options: .refreshCached,
                                             progress:  { (receivedSize, expectedSize, targetUrl) -> Void in
                                                self.delegate?.progress?(receivedSize:receivedSize, expectedSize:expectedSize)
        },
                                             completed: { (image, data, error, cacheType, finished, imageURL) -> Void in
                                                if let image = image {
                                                    self.rightImage = image
                                                    self.processImage()
                                                } else {
                                                    self.delegate?.failure()
                                                }
        })
    }
    
    func processImage() {
        if let leftImage = self.leftImage, let rightImage = self.rightImage {
            let anaglyphImage = ImageUtility.anaglyph(left: leftImage, right:rightImage)
            self.delegate?.finished(image:anaglyphImage)
            self.leftImage = nil
            self.rightImage = nil
            self.loadInProgress = false
        }
    }
    
    func fieldOfView() -> Double {
        let imageId = Mission.imageId(url: sourceUrl)
        let cameraId = Mission.currentMission().getCameraId(imageId: imageId)
        return Mission.currentMission().getCameraFOV(cameraId: cameraId) 
    }
    
    func imageId() -> String {
        return Mission.imageId(url: sourceUrl)
    }
    
    func angularDistance(otherImage:MarsPhoto) -> Double {
        guard modelJson != nil && otherImage.modelJson != nil else {
            return 0.0
        }
        
        let v1 = CameraModelUtils.pointingVector(self.modelJson!)
        let v2 = CameraModelUtils.pointingVector(otherImage.modelJson!)
        let dotProduct = dot(v1,v2)
        return acos(dotProduct)
    }
}
